clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -342,6 +346,17 @@ static bool isResourceRecordTypeOrArrayOf(VarDecl *VD) {
   return Ty->isHLSLResourceRecord() || Ty->isHLSLResourceRecordArray();
 }
 
+static const HLSLAttributedResourceType *
+getResourceArrayHandleType(VarDecl *VD) {
+  assert(VD->getType()->isHLSLResourceRecordArray() &&
+         "expected array of resource records");
+  const Type *Ty = VD->getType()->getUnqualifiedDesugaredType();
+  while (const ConstantArrayType *CAT = dyn_cast<ConstantArrayType>(Ty)) {
+    Ty = CAT->getArrayElementTypeNoTypeQual()->getUnqualifiedDesugaredType();
+  }
```

Comment:
This is grabbing the last value in the CAT?

also nit:
```suggestion
  while (const ConstantArrayType *CAT = dyn_cast<ConstantArrayType>(Ty))
    Ty = CAT->getArrayElementTypeNoTypeQual()->getUnqualifiedDesugaredType();
```

Reply:
It is grabbing the array element type (=the actual resource type). Multi-dimensional arrays are represented by nested `ConstantArrayType`s instances, so to get to the element type the array type needs to be "unwrapped" in a loop.